docs/tpm2-pkcs11_object_auth_model.md

```diff
@@ -0,0 +1,528 @@
+operations like unseal, encryption etc. can be restricted to two separate
+authorizations models if needed. This is done when defining the object using
+attributes - userwithauth and adminwithpolicy.<br>
+<br>
+It is also possible to associate the authorization of one object as the
+authorization of another object. One caveat of changing authorization value of
+a secondary object is that it does not invalidate the old object, instead
+creates a new object that protects the same sensitive information. Also, while
+auth values can be altered the authorization policy cannot be changed.
+
+### The NV Index/ Object
+There are two types of TPM NV objects - those that are made persistent after
+creating/ importing the objects under a TPM hierarchy OR those defined as NV
+indexes to hold user defined data. One can define an NV index of size 0 with an
+authorization password and or policy for administrative or user operations.
+Unlike the secondary objects, the NV index object authorization is not used in
+the encryption of it's contents - the NV index object authorization simply
+protects read and write access to the object. Thus no new NV indexes need to be
+defined when changing the authorization.
+
+## tpm2-pkcs11 objects
+A general overview of the object and authorization model design with TPM2
+objects. The SO and USER roles separation is realized in the tpm2-pkcs11 design
+by two unique TPM2 objects, either transient objects or NV indexes. The
+authorization values chosen for these objects are further utilized in creating
+specialized authorization policies that render the special properties of the
+SO-PIN, USER-PIN, Wrapping-Key-Sealing and the USER(keys) objects.
+1. SO-PIN object: The auth value of this object authorizes auth value change of
+SO-PIN and USER-PIN objects. This is either an NV index or a transient secondary
+object.
+2. USER-PIN object: The auth value of this object is used as an auth value of
+USER-PIN, Wrapping-Key-Sealing, and USER(keys) objects. The USER-PIN object auth
+value OR SO-PIN object auth value can authorize change of USER-PIN object auth
+value. This is either an NV index or a transient secondary object.
+3. Wrapping-Key-Sealing object: The authorization value of this object is
+referenced off the USER-PIN object with a policy. The object itself doesn't have
+a auth value of it's own and so the unsealing user operation would need the auth
+policy requiring USER pin ought to be satisfied. It seals a wrapping key used to
+encrypt the auth value of existing TPM2 objects to be assumed into the
+tpm2-pkcs11 model. This is a transient secondary object.
+4. USER(keys) object: The authorization model is same as the
+Wrapping-Key-Sealing object and thus requires the USER-PIN object auth value via
+a policy. These are user created keys and objects that are used in user
+operations like encryption, decryption, signing, verification, etc. These are
+secondary transient objects.
```

Comment:
We should define public/certificate vs private here and make sure early on we have tests for this. We probably can just skip the TPM for all these and just use OSSL directly.